toolchain/check/handle_name.cpp

```diff
@@ -168,6 +173,30 @@ auto HandleParseNode(Context& context, Parse::NameQualifierId /*node_id*/)
   return true;
 }
 
+auto HandleParseNode(Context& context, Parse::DesignatorExprId node_id)
+    -> bool {
+  SemIR::NameId name_id = context.node_stack().PopName();
+
+  // TODO: set `required` to `false` and check for not-found to give a clearer
+  // error when .Self is not found.
+  // FIXME: should this be a call to HandleNameAsExpr instead?
```

Comment:
That might be the easiest way to get the constant & NameRef structure (I do think you want that, though @zygoloid would probably be faster at coming up with test cases). I don't know if you want a helper as noted below to avoid a push/pop in the member access case, I think either approach might be fine.

Reply:
In principle, yes, this should be using `HandleNameAsExpr` I think, though I don't think it really makes much difference right now:

The work to map to a value in a specific is probably not going to be very important here -- that mostly affects cases where we find an unqualified lookup result in an extended scope such as a base class, and I don't think we expect to find `.Self` in such a scope -- but it wouldn't be wrong to do it, and it'd be more correct under maintenance / future language changes.

Building a `NameRef` would be nice (for consistency and for tooling if nothing else), though `.Self` is a bit weird compared to other names we might reference because it's not explicitly declared. I think I'd still be slightly on the side of including a `NameRef` though -- and more generally calling `HandleNameAsExpr`.

In terms of dealing with the undesired push, maybe the cleanest thing would be to move the existing push out to the callers. It'd be repeated three times, but I think it's most obvious what effect a handler has on the stack if it directly does all of its pushes and pops, so that seems OK to me.

Reply:
> Building a `NameRef` would be nice (for consistency and for tooling if nothing else), though `.Self` is a bit weird compared to other names we might reference because it's not explicitly declared.

I think `self` and `Self` are similar, and using `HandleNameAsExpr`.

https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/check/handle_name.cpp#L150-L163

Reply:
Should I only call `HandleNameAsExpr` in the `name_id == SemIR::NameId::SelfValue` / `.Self` branch of the `if` and do something else in the `.Member` branch?

Reply:
In particular, do we want to have a `.Member` produce a name reference to `.Self`?

Reply:
I've updated the PR so that it calls `HandleNameAsExpr` in both branches of the `if`, and so always produces a name reference to `.Self`. I added a note in the `.Member` case to explain the error when `.Selef` is not found in the current scope.